Extension/src/common.ts

```diff
@@ -848,6 +849,96 @@ export function isExecutable(file: string): Promise<boolean> {
     return pathAccessible(file, fs.constants.X_OK);
 }
 
+/** When on windows, ensures that the given executable name ends in an '.exe'.
+ * @param executableName The name of the executable to check.
+ * @returns The executable name with .exe appended if on windows and the name does not already end in an '.exe'.
+ */
+export function executableName(executableName: string) {
+    return isWindows && !/\.exe$/i.test(executableName) ? `${executableName}.exe` : executableName;
+}
+
+/** Returns true if the path is a folder.
+ *
+ * @param path The path to check.
+ * @returns A promise that resolves to true if the path is a folder, false otherwise.
+*/
+export async function isFolder(path: string) {
+    try {
+        return (await stat(path)).isDirectory();
+    } catch {
+        //
+    }
+    return false;
+}
+
+/** Recursively searches for all the files that match the filename (string or RegExp) in a given folder, up to a maxiumum depth.
+ *
+ * This will parallelize the search for all files in the folder and its subfolders.
+ *
+ * @param folder The folder to search in.
+ * @param filename The filename to search for (string or RegExp).
+ * @param predicate A function that takes a binary file path and returns a boolean indicating whether to include it in the results.
+ * @param maxDepth The maximum depth to search (default is 4).
+ * @param results An array to store the results.
+ * @returns A promise that resolves to an array of file paths that match the filename.
+ */
+export async function searchFolder(folder: string, filename: string | RegExp, predicate?: (binary: string) => Promise<boolean>, maxDepth = 4, results = new Array<string>()): Promise<string[]> {
+    try {
+        const files: string[] = await readdir(folder);
+        const all = files.map(async (file) => {
+
+            const fullPath: string = path.resolve(folder, file);
```

Comment:
Can it avoid processing the same directory multiple times from symlinks? I'm seeing stuff like /usr/bin/X11/X11/X11/X11/X11/X11/X11/time because X11 is symlink to `.`. The depth prevents it from being infinite, but it seems like unnecessary processing. Should something like a visited set of "realpath" (resolved paths) be used to avoid duplicate processing?